services/background-worker/src/Wire/BackendNotificationPusher.hs

```diff
@@ -0,0 +1,90 @@
+import Wire.API.Federation.BackendNotifications
+import Wire.API.Federation.Client
+import Wire.BackgroundWorker.Env
+
+startPushingNotifications ::
+  Q.Channel ->
+  Domain ->
+  AppT IO Q.ConsumerTag
+startPushingNotifications chan domain = do
+  lift $ ensureQueue chan domain
+  QL.consumeMsgs chan (routingKey domain) Q.Ack (pushNotification domain)
+
+-- | This class exists to help with testing, making the envelope in unit test is
+-- too difficult. So we use fake envelopes in the unit tests.
+class RabbitMQEnvelope e where
+  ack :: e -> IO ()
+  reject :: e -> Bool -> IO ()
+
+instance RabbitMQEnvelope Q.Envelope where
+  ack = Q.ackEnv
+  reject = Q.rejectEnv
+
+pushNotification :: RabbitMQEnvelope e => Domain -> (Q.Message, e) -> AppT IO ()
+pushNotification targetDomain (msg, envelope) = do
+  -- Jittered exponential backoff with 10ms as starting delay and
+  -- 300s as max delay.
+  --
+  -- FUTUREWORK: Pull these numbers into config
+  let policy = capDelay 300_000_000 $ fullJitterBackoff 10000
+      logErrr willRetry (SomeException e) rs =
+        Log.err $
+          Log.msg (Log.val "Exception occurred while pushing notification")
+            . Log.field "error" (displayException e)
+            . Log.field "domain" (domainText targetDomain)
+            . Log.field "willRetry" willRetry
+            . Log.field "retryCount" rs.rsIterNumber
+      skipChanThreadKilled _ = Handler $ \(_ :: Q.ChanThreadKilledException) -> pure False
+      handlers =
+        skipAsyncExceptions
+          <> [ skipChanThreadKilled,
+               logRetries (const $ pure True) logErrr
+             ]
+  recovering policy handlers $ const go
+  where
+    go :: AppT IO ()
+    go = case A.eitherDecode @BackendNotification (Q.msgBody msg) of
+      Left e -> do
+        Log.err $
+          Log.msg (Log.val "Failed to parse notification, the notification will be ignored")
+            . Log.field "domain" (domainText targetDomain)
+            . Log.field "error" e
+
+        -- FUTUREWORK: This rejects the message without any requeueing. This is
+        -- dangerous as it could happen that a new type of notification is
+        -- introduced and an old instance of this worker is running, in which case
+        -- the notification will just get dropped. On the other hand not dropping
+        -- this message blocks the whole queue. Perhaps there is a better way to
+        -- deal with this.
+        lift $ reject envelope False
+      Right notif -> do
+        ceFederator <- asks federatorInternal
+        ceHttp2Manager <- asks http2Manager
+        let ceOriginDomain = notif.ownDomain
+            ceTargetDomain = targetDomain
+            fcEnv = FederatorClientEnv {..}
+        liftIO $ either throwM pure =<< sendNotification fcEnv notif.targetComponent notif.path notif.body
+        lift $ ack envelope
+
+-- FUTUREWORK: Recosider using 1 channel for many consumers. It shouldn't matter
+-- for a handful of remote domains.
```

Comment:
The channel itself doesn't block when a domain is unreachable. It only blocks the particular consumer who is supposed to be pushing notifications to the unreachable domain. Other consumers keep going.

I suspect that this will reach some limit of consumers or just be slow when we have hundreds of consumers running. But perhaps we can solve that problem when we get there.